File: 19. Counting Sundays/counting_sundays.py
import time
import datetime as dt


class Calendar1:
    def __init__(self, date1, date2):
        self.d1 = dt.date(*[int(i) for i in date1.split('-')])
        self.d2 = dt.date(*[int(i) for i in date2.split('-')])
    
    def _get_next_month(self, date):
        cur_mth = date.month
        cur_yr = (date.year) + cur_mth//12
        cur_mth = (cur_mth ) % 12  + 1
        return dt.date(cur_yr, cur_mth, date.day)

    def solve(self):
        count = 0
        while self.d1 <= self.d2:
            if self.d1.weekday() == 6:
                count += 1
            self.d1 = self._get_next_month(self.d1)
        return count



class Calendar2:
    def __init__(self, start_date, end_date):
        self.date = self._to_dt('1900-01-01')
        self.date1 = self._to_dt(start_date)
        self.date2 = self._to_dt(end_date)
        self.count = 0
        self.days_diff = 0
    
    def _to_dt(self, date):
        self._format = '%Y-%m-%d'
        return dt.datetime.strptime(date, self._format)

    def _calculate_days(self, d):
        return (d - self.date).days
    
    def _check_day(self, d):
        self.days_diff = self._calculate_days(d)
        return (self.days_diff + 1) % 7
    
    def _get_next_month(self, date):
        cur_mth = date.month
        cur_yr = (date.year) + cur_mth//12
        cur_mth = (cur_mth ) % 12  + 1
        return dt.datetime(cur_yr, cur_mth, date.day)

    def solve(self):  
        cur_month = self.date1
        while cur_month < self.date2:
            # print(self.count, cur_month)
            if self._check_day(cur_month) == 0:
                self.count += 1
                # print(self.date1)
            cur_month = self._get_next_month(cur_month)
        return self.count
    


print('Calendar 1 : ', Calendar1('1901-01-01', '2000-12-31').solve())
print('Calendar 2 : ', Calendar2('1901-01-01', '2000-12-31').solve())

File: 19. Counting Sundays/counting_sundays.js
class Calendar{
    constructor(date1, date2){
        this.date1 = new Date(date1);
        this.date2 = new Date(date2);
        this.count = 0
    }

    // check given year is leap year or not
    isLeapYear(date){
        var year = date.getFullYear();
        if(year%4 == 0){
            if(year % 100 == 0){
                if(year % 400 == 0){
                    return true;
                } else{
                    return false;
                }
            } else{
                return true;
            }
        } else{
            return false;
        }
    }

    getMonthDays(date){
        var feb_days =  date.getFullYear() ? 29 : 28;     
        var days = {1:31, 2:feb_days, 3:31, 5:31, 7:31, 8:31, 10:31, 12:31, 4: 30, 6: 30, 9: 30, 11: 30 }
        // console.log(date.getMonth()+1);
        return days[date.getMonth()+1]; 
    }

    getNextMonth(){
        if(this.date1.getMonth() == 12){
            this.date1.setYear(this.date1.getFullYear()+1);
            this.date1.setMonth(1);            
        } else{
            this.date1.setMonth(this.date1.getMonth()+1);
        }
    }

    calculateDays(){
        var days = 0;
        // console.log(date.getMonth());

        while(this.date1 < this.date2){
            days += this.getMonthDays(this.date1);
            var rem = 0;
            rem = (rem + days)%7;
            if(rem == 0){
                this.count++;
                // console.log(this.date1);
            }
            this.getNextMonth();
        }
        return this.count;
    }
}

var a = new Calendar('1901-01-01', '2000-12-01');
console.log(a.calculateDays());
